helpers/target.go

```diff
@@ -426,6 +458,57 @@ func IsAWSAccReachable(accARN, assumeRoleURL, role string, sessDuration int) (bo
 		assumeRoleResp.SessionToken), nil
 }
 
+// IsGCPProjReachable returns whether the GCP project is reachable.
+//
+// If the environment variable VULCAN_SKIP_REACHABILITY is true
+// according to [strconv.ParseBool], then the reachability test is
+// skipped and IsGCPProjReachable returns true.
+func IsGCPProjReachable(gcpProject, endpoint, saCreds string) (bool, error) {
+	if skipReachability() {
+		return true, nil
+	}
+
+	if !types.IsGCPProjectID(gcpProject) {
+		return false, nil
+	}
+
+	saCredsJSON := []byte(saCreds)
+
+	ctx := context.Background()
+	var serviceOptions []option.ClientOption
+	if endpoint != "" {
+		serviceOptions = append(serviceOptions, option.WithEndpoint(endpoint))
+		serviceOptions = append(serviceOptions, option.WithoutAuthentication())
+	} else {
+		serviceOptions = append(serviceOptions, option.WithCredentialsJSON(saCredsJSON))
+	}
+	crmService, err := cloudresourcemanager.NewService(ctx, serviceOptions...)
```

Comment:
I think, if you pass cloudresourcemanager as an argument to the function, you could, for example, test the behaviour when NewService returns an error because it is easier to mock in this way.

Reply:
I was doing that initially, however, it was not aligned with `ServiceCreds` interface implementation, so I went the other way around. if you have a good reference implementation for the same, please let me know :)

Reply:
What about something like that:
In the method IsReachable, you can put:

```go
	case gcpProjType:
		ps, err := newProjectService(creds.URL(), creds.Password())
		if err != nil {
			return isReachable, err
		}
		isReachable, err = IsGCPProjReachable(target, ps)
```

The implementation of newProjectService would be:

```go
func newProjectService(endpoint, saCreds string) (*crm.ProjectsService, error) {
	saCredsJSON := []byte(saCreds)

	ctx := context.Background()
	var serviceOptions []option.ClientOption
	if endpoint != "" {
		serviceOptions = append(serviceOptions, option.WithEndpoint(endpoint))
		serviceOptions = append(serviceOptions, option.WithoutAuthentication())
	} else {
		serviceOptions = append(serviceOptions, option.WithCredentialsJSON(saCredsJSON))
	}
	crmService, err := crm.NewService(ctx, serviceOptions...)
	if err != nil {
		return nil, err
	}
	projectService := crm.NewProjectsService(crmService)

	return projectService, nil
}
```

And finally the function IsGCPProjReachable will be:

```go
func IsGCPProjReachable(gcpProject string, ps *crm.ProjectsService) (bool, error) {
	if skipReachability() {
		return true, nil
	}

	if !types.IsGCPProjectID(gcpProject) {
		return false, nil
	}

	projectName := fmt.Sprintf("projects/%s", gcpProject)
	projectService := ps.Get(projectName)
	project, err := projectService.Do()
	if err != nil {
		// Parsing Google API error to check if the error is due to a forbidden access.
		var e *googleapi.Error
		if errors.As(err, &e) {
			if e.Code == http.StatusForbidden {
				return false, nil
			}
		}
		return false, err
	}

	if project.State != "ACTIVE" {
		return false, nil
	}

	return true, nil
}
```

I think, in this way will be easier to test IsGCPProjReachable mocking crm.ProjectsService.


Note that I changed the import of the `cloudresourcemanager` because it was too long and I used `crm "google.golang.org/api/cloudresourcemanager/v3"` instead.

Reply:
I've been thinking about all of this, and for now, we are going to leave it as you did it because it's too much effort only to change the way to test it. Also, we maintain the uniformity with the AWS case